File: buildTree.java
import java.util.*;
// Approach 1- DFS (Recursion) [using implicit call-stack]
class Solution { // TC O(n) SC O(n)
  int post_idx;
  int[] postorder;
  int[] inorder;
  HashMap<Integer, Integer> idx_map = new HashMap<Integer, Integer>();

  public Node helper(int in_left, int in_right) {
    // if there is no elements to construct subtrees
    if (in_left > in_right)
      return null;

    // pick up post_idx element as a root
    int root_val = postorder[post_idx];
    Node root = new Node(root_val);

    // root splits inorder list
    // into left and right subtrees
    int index = idx_map.get(root_val);

    // recursion 
    post_idx--;
    // build right subtree
    root.right = helper(index + 1, in_right);
    // build left subtree
    root.left = helper(in_left, index - 1);
    return root;
  }

  public Node buildTree(int[] inorder, int[] postorder) {
    this.postorder = postorder;
    this.inorder = inorder;
    // start from the last postorder element
    post_idx = postorder.length - 1;

    // build a hashmap value -> its index
    int idx = 0;
    for (Integer val : inorder)
      idx_map.put(val, idx++);
    return helper(0, inorder.length - 1);
  }
}

class buildTree{
  public static void main(String[] args) {
    int[] inorder={9,3,15,20,7};
    int[] postorder={9,15,7,20,3};
    Solution s= new Solution();
    BinaryTree bt=new BinaryTree();
    Node result=s.buildTree(inorder, postorder);
    bt.preorder(result);
  }
}

File: buildTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    def preorder(self, root):
        if not root: return
        print(root.val, end=" ")
        self.preorder(root.left)
        self.preorder(root.right)

class Solution:
# Approach 1- DFS (Recursion) [using implicit call-stack]
    def buildTree(self, inorder, postorder): # TC O(n) // SC O(n)
        def helper(in_left, in_right):
            # if there is no elements to construct subtrees
            if in_left > in_right:
                return None
            
            # pick up the last element as a root
            val = postorder.pop()
            root = TreeNode(val)

            # root splits inorder list
            # into left and right subtrees
            index = idx_map[val]
 
            # build right subtree
            root.right = helper(index + 1, in_right)
            # build left subtree
            root.left = helper(in_left, index - 1)
            return root
        
        # build a hashmap value -> its index
        idx_map = {val:idx for idx, val in enumerate(inorder)} 
        return helper(0, len(inorder) - 1)

s=Solution()
result=(s.buildTree([9,3,15,20,7],[9,15,7,20,3]))
t=TreeNode()
print(t.preorder(result))
